[Lev1][미션3] 블랙잭 1단계 - 주노

Choi-JJunho opened the pull request:
잘부탁드립니다..^^

Songusika committed:
docs: .gitignore 수정

Co-authored-by: choi-jjunho <junho5336@gmail.com>

Songusika committed:
docs: 기능목록 정의

Co-authored-by: choi-jjunho <junho5336@gmail.com>

Songusika committed:
feat: 카드 도메인 기능 구현

Co-authored-by: choi-jjunho <junho5336@gmail.com>

Songusika committed:
refactor: 카드 숫자, 모양 분리

Co-authored-by: choi-jjunho <junho5336@gmail.com>

Songusika committed:
feat: 카드팩 생성기능 추가

Co-authored-by: choi-jjunho <junho5336@gmail.com>

Songusika committed:
refactor: 카드팩 카드반환 기능 추가

Co-authored-by: choi-jjunho <junho5336@gmail.com>

Songusika committed:
feat: 사용자 기능 구현

Co-authored-by: choi-jjunho <junho5336@gmail.com>

Songusika committed:
feat: 심판 기능 구현

Co-authored-by: choi-jjunho <junho5336@gmail.com>

Songusika committed:
mission: 제네릭 미션 추가

Co-authored-by: choi-jjunho <junho5336@gmail.com>

Songusika committed:
feat: 입력 뷰 구현

Co-authored-by:f choi-jjunho <junho5336@gmail.com>

Songusika committed:
feat: 블랙잭 게임 구현

Co-authored-by: choi-jjunho <junho5336@gmail.com>

Songusika committed:
refactor: 코드 리포맷팅

Co-authored-by: choi-jjunho <junho5336@gmail.com>

Songusika committed:
docs: README.md 업데이트

Co-authored-by: choi-jjunho <junho5336@gmail.com>

Songusika committed:
refactor: Score -> GameResult 클래스 명 변경

Co-authored-by: choi-jjunho <junho5336@gmail.com>

greeng00se commented on README.md:
```diff
@@ -5,3 +5,82 @@
+    - [x] `,`로 각 플레이어를 구분한다.
+- [x] 각 플레이어의 진행 여부를 입력받는다.
+
+### 👀️ 출력 뷰
```
이모지로 인해 가독성이 더 좋아진 것 같아! 읽는데 재밌는 느낌?

be-student commented on .gitignore:
```diff
@@ -30,3 +30,6 @@ out/
 .vscode/
+
+### git Template ###
+.gitmessage.txt
```
오 이거 좋네요 알아갑니다

Choi-JJunho replied:
https://velog.io/@junho5336/%ED%8E%98%EC%96%B4%ED%94%84%EB%A1%9C%EA%B7%B8%EB%9E%98%EB%B0%8D%EC%9D%B8%EB%8D%B0-%EC%BB%A4%EB%B0%8B%EC%9D%80-%ED%95%9C%EB%AA%85%EC%9D%B4

greeng00se commented on src/main/java/blackjack/controller/BlackjackController.java:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    private void processPlayerDraw(final BlackjackGame blackjackGame, final Player player) {
+        while (inputView.readCommand(player.getName()) && !isBust(player.showCards())) {
```
isBust 메서드가 클린코드 p.46 페이지에 있는 내려가기 규칙을 적용했으면 더 읽기 편했을 것 같아!

greeng00se commented on src/main/java/blackjack/controller/BlackjackController.java:
```diff
@@ -0,0 +1,103 @@
+
+    private void startGame(final Players players, final Dealer dealer, final BlackjackGame blackjackGame) {
+        blackjackGame.initGame(dealer);
+        players.getPlayers().forEach(blackjackGame::initGame);
```
for-each를 사용하지 않고 collection의 forEach를 사용한 이유가 있을까? (단순 궁금)

Choi-JJunho replied:
페어가 좋아하는 방식이라 이번에collection의 foreach를 사용해봤어!

만약 사용 근거를 생각해본다면 문장의 축약이 이유가 될것 같네

greeng00se commented on src/main/java/blackjack/controller/BlackjackController.java:
```diff
@@ -0,0 +1,103 @@
+        players.getPlayers().forEach((player) -> {
+            final GameResult score = referee.askResultByUserName(player.getName());
+            result.put(player.getName(), score);
+        });
```
hashMap을 사용한다면 다음과 같이 축약할 수 있을 것 같아 👍 사실 안돌려봐서 맞는지는 모르겠다 😄 
```
return players.getPlayers().stream()
        .map(Player::getName)
        .collect(player -> toMap(player, refree.askResultByUserName(player)));
```

greeng00se commented on src/main/java/blackjack/domain/cardpack/CardPack.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    public Card pop() {
+        return cards.remove(cards.size() - 1);
```
pop보다 블랙잭과 관련된 용어를 사용해보면 어떨까? pop은 뭔가 자료구조에 대한 메서드명 느낌인 것 같아

greeng00se commented on src/main/java/blackjack/domain/user/Players.java:
```diff
@@ -0,0 +1,39 @@
+        validateDuplicatedNames(players);
+    }
+
+    private void validatePlayerCount(final List<Player> players) {
+        if (players.size() > MAX_PLAYER_COUNT) {
+            throw new IllegalArgumentException("플레이어들은 13명 이하여야 합니다.");
+        }
+    }
```
딜러까지 총 14명인데 52장의 카드로 부족할 수도 있지 않을까?
플레이어1: 1+1+1+1+2+2+2+2+3+3+10 28
플레이어2: 3+3+4+4+4+10 = 28
플레이어3: 4+5+5+5+10 = 29
플레이어4: 5+6+6+10 + 27
플레이어5: 6+6+7+10 = 29
플레이어6: 7+7+7 = 21
플레이어7: 8+8+8 = 24
플레이어8: 8+9+9 = 26
플레이어9: 9+9+10 = 28
플레이어10: 10+10+10 = 30
플레이어11: 10+10+10 = 30

위 예시는 극단적인 상황이지만 말이지..

Choi-JJunho replied:
헉.. 그러네!!?
수정해볼게👍

greeng00se commented on src/main/java/blackjack/domain/user/name/PlayerName.java:
```diff
@@ -0,0 +1,37 @@
+        validateReservedWord(candidateName);
+    }
+
+    private static void validateReservedWord(final String name) {
+        if ("딜러".equals(name)) {
+            throw new IllegalArgumentException("사용자 이름은 딜러일 수 없습니다.");
+        }
+    }
+
+    private static void validateLength(final String name) {
+        if (name.length() > 10) {
+            throw new IllegalArgumentException("사용자 이름은 10자 이하여야 합니다.");
+        }
+    }
```
딜러와 10은 상수화해주는건 어때?

Choi-JJunho replied:
미처 못봤던 부분이네👍

greeng00se commented on src/main/java/blackjack/view/Command.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    private boolean equalsByQuery(final String query) {
+        return command.equalsIgnoreCase(query);
```
equalsIgnoreCase 좋은데요? 👍

Choi-JJunho replied:
대문자 Y, N도 올바른 입력으로 고려해서 사용했어 👍

greeng00se reviewed (commented):
README에 이모지 있으니까 산뜻한데? 리뷰 작게 남기고 갑니당~

be-student commented on src/test/java/blackjack/domain/user/name/UserNameTest.java:
```diff
@@ -0,0 +1,19 @@
+        UserName userName = new UserName(" 이름이에요 ");
+
+        // then
+        Assertions.assertThat(userName.getName()).isEqualTo("이름이에요");
```
```java
 Assertions.assertThat(userName.getName())
.isEqualTo("이름이에요");
```
같은 형태로 개행 해주는 것이 좋다고 하네요

Choi-JJunho replied:
IntelliJ 라인 전체정렬하면서 개행이 망가진게 몇개 있는거같아ㅠㅠ

개행은 한번 전체적으로 검토해봐야겠네

be-student commented on src/main/java/blackjack/domain/user/name/DealerName.java:
```diff
@@ -0,0 +1,8 @@
+public class DealerName extends UserName {
+
+    public DealerName() {
+        super("딜러");
```
단순히 궁금한 건데, new UserName("딜러") 를 사용하지 않은 이유는 뭐야?

Choi-JJunho replied:
`딜러는 고정된 이름을 가진다`라는 기준을 잡고 DealerName의 생성자에 해당 값을 고정하는 형태로 작성했어!

be-student commented on src/main/java/blackjack/view/OutputView.java:
```diff
@@ -0,0 +1,90 @@
+
+        printUserCards(dealer);
+        System.out.println();
+        players.getPlayers().forEach(player -> {
+            printUserCards(player);
+            System.out.println();
+        });
```
메서드에서 개행까지 해주는 거는 어떨까요?

Choi-JJunho replied:
해당 메소드를 다른 메소드에서도 같이 사용해서 개행을 밖으로 따로 분리해줬어!

be-student commented on src/main/java/blackjack/domain/user/Players.java:
```diff
@@ -0,0 +1,39 @@
+
+    private void validateDuplicatedNames(final List<Player> players) {
+
+        final long count = players.stream().map(User::getName).distinct().count();
```
stream 에서 .마다 줄바꿈 해주면 좋을 것 같아요

Choi-JJunho replied:
이부분도 전체정렬하면서 놓쳤나봐😂

be-student commented on src/main/java/blackjack/domain/game/ResultReferee.java:
```diff
@@ -0,0 +1,52 @@
+public class ResultReferee {
+
+    private final ScoreBoard scoreBoard;
+    private final Map<GameResult, Integer> dealerScoreBoard = new EnumMap<>(GameResult.class);
```
요구사항에 일급 컬렉션을 사용하라고 적혀있으니, 인스턴스 변수의 수를 줄여도 괜찮을 것 같아요

be-student commented on src/main/java/blackjack/domain/game/BlackjackGame.java:
```diff
@@ -0,0 +1,43 @@
+
+    public boolean isDealerEnd(Dealer dealer) {
+        final int dealerScore = ScoreReferee.calculateScore(dealer.showCards());
+        return (dealerScore > MIN_DEALER_SCORE) || (dealerScore == BUST_SCORE);
```
여기 메서드의 느낌이 dealerScore > BUST_SCORE 이라는 경우가 안 보이는 느낌이 들어서 dealerScore >=BUST_SCORE 같은 느낌으로 해보는건 어떤가요?

Choi-JJunho replied:
`딜러는 16점을 넘거나 버스트 점수(-1)가 되면 종료된다`
가 딜러의 종료 조건으로 설정했는데 메소드 분리를 생각해봐야겠네👍

be-student commented on src/main/java/blackjack/domain/game/ScoreBoard.java:
```diff
@@ -0,0 +1,50 @@
+        scoreBoard.put(user, score);
+    }
+
+    private User getUserByName(final UserName queryName) {
+        return scoreBoard.keySet().stream()
+                .filter(user -> user.getName().equals(queryName))
+                .findFirst()
+                .orElseThrow(() -> new IllegalArgumentException("해당 이름을 찾을 수 없습니다."));
+    }
```
Map을 사용했는데, userName을 통해서 검색할 수 없다는 것은 살짝 이상할 것 같아요
UserName 을 바탕으로 키를 설정해보면 어떨까요?

Choi-JJunho replied:
처음에는 누누 말대로 UserName을 바탕으로 key를 구성했는데 ScoreBoard의 역할 중 `점수 기록`이라는 역할을 수행하는 과정에서 각 플레이어들의 손패가 필요하게 되어 위와같이 `손패를 가지고있는 플레이어`를 키값으로 구성하게 되었어 😂

역할의 분리에 조금 더 신경써봐야겠다 👍

be-student commented on src/main/java/blackjack/view/OutputView.java:
```diff
@@ -0,0 +1,90 @@
+
+public class OutputView {
+
+    public void printUsersCard(final Dealer dealer, final Players players) {
```
도메인의 객체들이 그대로 나가는 느낌이라, DTO 같은 느낌으로 연결을 끊어보시는 것은 어떤가요?

be-student reviewed (commented):
진짜 잘 봤어 주노 너무 코드 깔끔하네

drunkenhw commented on README.md:
```diff
@@ -5,3 +5,82 @@
+
+### 🃏🃏 도메인 - 카드팩
+
+- [x] 카드팩은 생성 시 순서를 가진다.
```
순서가 가진다? 왜 이런 요구사항이 필요하게 된 거죠?

Choi-JJunho replied:
카드팩을 처음 샀을 때 순서대로 정렬되어있는 모습을 생각하며 구상하여 위와같은 요구사항이 나왔네요🤔

지금보니 도메인을 너무 현실에 빗대어 표현한건 아닌가라는 생걱이 들기도 하네요

drunkenhw commented on src/main/java/blackjack/controller/BlackjackController.java:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    private boolean isBust(final List<Card> showCards) {
+        return ScoreReferee.calculateScore(showCards) == BUST_SCORE;
```
`ScoreReferee`한테 Bust 인지 물어보는 건 어떤가요? 그리고 이 메소드는 Controller에 있어야할까요? BlackjackGame에 있어도 좋을 것 같은데요?

Choi-JJunho replied:
Game의 역할이 맞아보이네요 👍

drunkenhw commented on src/main/java/blackjack/domain/cardpack/MasterShuffleStrategy.java:
```diff
@@ -0,0 +1,13 @@
+import java.util.Collections;
+import java.util.List;
+
+public class MasterShuffleStrategy implements ShuffleStrategy {
```
MasterShuffle은 무슨 뜻인가요?

hanueleee replied:
나두 궁금. 기본 셔플이라는 뜻인가??

Choi-JJunho replied:
페어와 막 생각한 이름이였는데
`숙련자의 shuffle`이란 뉘앙스로 정말로 랜덤하게 섞인다는 것을 표현하고 싶었어 😂

drunkenhw commented on src/main/java/blackjack/domain/game/ResultReferee.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    public Map<GameResult, Integer> getDealerScore() {
+        return new EnumMap<>(dealerScoreBoard);
```
`List.copyOf()`나 `unmodifiableList()` 를 안쓰고 `new`를 사용하신 이유가 있을까요?

Choi-JJunho replied:
특별한 이유는 없습니다

자바 11이랑 아직 덜친해져서 그랬던거같네요

drunkenhw commented on src/main/java/blackjack/domain/game/ScoreReferee.java:
```diff
@@ -0,0 +1,47 @@
+
+    private static List<Card> getAcePack(final List<Card> cards) {
+        return cards.stream()
+                .filter(card -> CardNumber.ACE.equals(card.getNumber()))
```
card 한테 isAce()냐고 물어보는 건 어때요?

Choi-JJunho replied:
좋은데요?

drunkenhw commented on src/main/java/blackjack/domain/user/name/PlayerName.java:
```diff
@@ -0,0 +1,37 @@
+package blackjack.domain.user.name;
+
+public class PlayerName extends UserName {
```
Name의 공통로직은 getName() 뿐인데 상속을 하신 이유가 뭔가요?

Choi-JJunho replied:
`유저(플레이어와 딜러)는 동일한 getName을 가지되 플레이어 이름과 딜러의 이름은 다른 검증과정을 거친다.`라는 발상에서부터 상속을 떠올리게 되어 위와같이 분리해봤습니다

drunkenhw commented on src/main/java/blackjack/domain/user/Player.java:
```diff
@@ -0,0 +1,10 @@
+
+import blackjack.domain.user.name.PlayerName;
+
+public class Player extends User {
```
`Dealer` 와 `Player`는 `User`를 상속받고 있는데요. 딜러, 플레이어에 아무 로직이 없는데 이런 경우에도 굳이 상속을 했어야할까요? `User` 하나로 처리할 수 있을 것 같은데요?

hanueleee replied:
박스터 의견에 동의합니다. 공통로직을 user에 묶고 개별로직을 나누어줄게 아니라면, user 하나로 처리해도 될 것 같습니다!

Choi-JJunho replied:
instance of를 통해 비즈니스 로직 진향간 플레이어와 딜러를 구분하기 위해 위와같이 구성했는데 방금 리뷰어에게 받은 리뷰로는 해당 방법이 OCP를 위반할 수도 있다고하네

인터페이스로 구성하는것도 하나의 방법일것같기도하네🤔

drunkenhw replied:
두 클래스간에 구분되는 역할이 없다면 `딜러`라는 이름으로 구분을 할 수 있지않을까요? 굳이 클래스를 안나누고 `instanceOf`를 사용하지 않아도 되는거 아닌가요?

Choi-JJunho replied:
죄송합니다..

drunkenhw commented on src/main/java/blackjack/view/InputView.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    public boolean readCommand(final UserName userName) {
+        System.out.println(
```
`printf()`는 어떠신가요?

drunkenhw commented on src/main/java/blackjack/view/OutputView.java:
```diff
@@ -0,0 +1,90 @@
+        System.out.println("딜러: " + stringBuilder);
+
+        playerScore.keySet().forEach(userName -> {
+            System.out.println(userName.getName() + ": " + playerScore.get(userName).getView());
```
주노님은 `forEach()` 매니아신거 같은데 쓰는 곳과 안쓰는 곳의 기준이 있을까요?

drunkenhw commented on src/test/java/blackjack/domain/card/CardTest.java:
```diff
@@ -0,0 +1,33 @@
+        Card card = new Card(number, CardShape.DIAMOND);
+
+        // then
+        assertThat(card.getNumber()).isEqualTo(number);
```
여기는 왜 when이 빠져있죠?

drunkenhw commented on src/test/java/blackjack/domain/cardpack/CardPackTest.java:
```diff
@@ -0,0 +1,80 @@
+
+        @Test
+        void _순서대로_생성된다() {
+
```
개행?

Choi-JJunho replied:
죄송합니다..

drunkenhw commented on src/test/java/blackjack/domain/cardpack/CardPackTest.java:
```diff
@@ -0,0 +1,80 @@
+
+        //then
+        Card card = cardPack.pop();
+        Assertions.assertThat(card.getShape()).isEqualTo(CardShape.SPADE);
+        Assertions.assertThat(card.getNumber()).isEqualTo(CardNumber.ACE);
```
`assertAll()`을 사용해보는 것은 어떨까요? https://www.baeldung.com/junit5-assertall-vs-multiple-assertions

drunkenhw commented on src/test/java/blackjack/domain/game/ResultRefereeTest.java:
```diff
@@ -0,0 +1,88 @@
+class ResultRefereeTest {
+
+    @Test
+    void 스코어보드를_받아서_각플레이어의_승_패_무_를_결정한다() {
```
테스트가 너무 길어 읽기가 힘드네요 좋은 방법이 없을까요?

Choi-JJunho replied:
죄송합니다 해주세요

drunkenhw commented on src/test/java/blackjack/domain/game/ScoreBoardTest.java:
```diff
@@ -0,0 +1,47 @@
+
+        // then
+        Assertions.assertThatThrownBy(() -> scoreBoard.get(noParticipant))
+                .isInstanceOf(IllegalArgumentException.class);
```
메세지까지 같이 검증해보는 것은 어떤가요?

drunkenhw commented on src/test/java/blackjack/domain/user/PlayerTest.java:
```diff
@@ -0,0 +1,29 @@
+
+        // then
+        Assertions.assertThat(player.showCards())
+                .containsExactly(new Card(CardNumber.KING, CardShape.CLOVER));
```
여기서는 카드 순서까지 알아야 검증이 되는 것 같은데요? `카드를 뽑을 수 있다`에서 굳이 순서를 알아야할까요? 카드를 한장 뽑았다는 size만 검증해도 괜찮지 않을까요?

drunkenhw commented on src/test/java/blackjack/domain/util/NoviceShuffleStrategy.java:
```diff
@@ -0,0 +1,14 @@
+import blackjack.domain.cardpack.ShuffleStrategy;
+import java.util.List;
+
+public class NoviceShuffleStrategy implements ShuffleStrategy {
```
Novice는 뭐죠?

Choi-JJunho replied:
초심자라는 뜻입니다
죄송합니다.

drunkenhw reviewed (commented):
안녕하세요 주노~ 리뷰어 박스터입니다.
전반적으로 코드는 깔끔하네요 궁금한 점 위주로 코멘트 달았습니다~

hanueleee commented:
안뇽하세요 구경왔습니다

hanueleee commented on src/main/java/blackjack/Application.java:
```diff
@@ -0,0 +1,12 @@
+
+public class Application {
+    public static void main(String[] args) {
+        BlackjackController controller = new BlackjackController(new InputView(), new OutputView());
```
inputView와 outputView를 controller 외부에서 주입해주는 이유가 있나요??

Choi-JJunho replied:
Application 테스트를 고려해서 주입받는 방향으로 구성했어!

입 출력에 대한 테스트를 하기에는 아직 어려움이 있지만.. 향후 테스트를 하게 된다면 보다 수월한 구조가 될 것 같다는 그런생각이야!

Choi-JJunho replied:
키워드 : 명시적 의존성 / 숨은 의존성 

[오브젝트 책에 있음]

hanueleee commented on src/main/java/blackjack/domain/cardpack/CardPack.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    public CardPack(final List<Card> cards) {
+        this.cards = new ArrayList<>(cards);
```
복사해서 넣어주는 이유가 따로 있나요??

Choi-JJunho replied:
원본 List와 연결점을 해지하고싶어서 생성자에서 복사해서 초기화했어!

hanueleee commented on src/main/java/blackjack/domain/game/ResultReferee.java:
```diff
@@ -0,0 +1,52 @@
+        return declareScore(playerScore, dealerScore);
+    }
+
+    private GameResult declareScore(final int playerScore, final int dealerScore) {
```
메소드 명이 너무 좋아요. 정말로 referee가 score를 declare해주는 느낌!!

hanueleee commented on src/main/java/blackjack/view/OutputView.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    public void printDealerDraw() {
+        System.out.println("딜러는 16이하라 한장의 카드를 더 받았습니다.");
```
문구를 상수로 빼는건 어떨까요??

hanueleee reviewed (commented):
안뇽하세요 놀러왔습니다 ><

Choi-JJunho committed:
chore: 셔플전략 패키지 변경

Choi-JJunho committed:
style: 코드 재정렬

Choi-JJunho committed:
refactor: 메소드 분리

Choi-JJunho committed:
refactor: 상수 추출

Choi-JJunho committed:
fix: 입력 메시지 수정

Choi-JJunho committed:
refactor: 상속 구조를 조합 구조로 수정

- User를 상속받아서 사용했던 구조를 Participant를 필드로 가지는 조합 구조로 변경했습니다.

- 기존의 ScoreReferee가 수행하던 점수 계산 로직을 Hand 클래스에 위임했습니다.

- 플레이어 / 딜러로부터 즉시 점수를 얻을 수 있습니다.

- 딜러로부터 딜러의 승부 결과를 얻을 수 있습니다.

- ResultReferee가 수행하던 로직을 Controller에 위임했습니다
  - 이제 Controller가 플레이어의 승부 결과를 모아서 View에서 사용하도록 변환해줍니다.

- 잘못된 입력 시 재입력을 할 수 있습니다.

Choi-JJunho committed:
refactor: 코드 리포맷팅

Choi-JJunho committed:
docs: 기능목록 최신화

Choi-JJunho committed:
refactor: 클래스명 변경 Hand -> Cards

Choi-JJunho committed:
refactor: 리뷰내용 적용 및 구조 개선

- 플레이어에게 getResult 기능 추가
- 기존 isEnd의 역할을 Dealer에게 위임
- 점수를 표현하는 Score 객체 추가
- 코드 포맷팅

Choi-JJunho committed:
refactor: Dto 추가

Choi-JJunho committed:
feat: 배당률 추가

Choi-JJunho committed:
feat: 돈 객체 추가

Choi-JJunho committed:
feat: 돈 더하기, 곱하기 기능 추가

Choi-JJunho committed:
feat: 수익 계산 및 출력기능 추가
